trinity/utils/datastructures.py

```diff
@@ -582,34 +601,74 @@ def _mark_one_task_complete(self, task_id: TTaskID) -> Generator[TTaskID, None,
         self._unready.remove(task_id)
 
         # prune any completed tasks that are too old
-        self._prune(task_id)
+        self._prune_finished(task_id)
 
         # resolve tasks that depend on this task
         for depending_task_id in self._dependencies[task_id]:
             # we already know that this task is ready, so we only need to check completion
             if self._tasks[depending_task_id].is_complete:
                 yield depending_task_id
 
-    def _prune(self, task_id: TTaskID) -> None:
+    def _prune_finished(self, task_id: TTaskID) -> None:
         """
-        This prunes any data starting at the task completed at task_completion, and older.
+        This prunes any data starting more than _max_depth in history.
         It is called when the task becomes ready.
         """
-        # determine how far back to prune
-        finished_depth = self._depths[task_id]
-
-        prune_depth = finished_depth - self._max_depth
-        if prune_depth > self._oldest_depth:
+        try:
+            oldest_id = self._find_oldest_unpruned_task_id(task_id)
+        except ValidationError:
+            # No tasks are old enough to prune, can end immediately
+            return
+
+        root_id, depth = self._find_root(oldest_id)
+        unpruned = self._prune_forward(root_id, depth)
+        if oldest_id not in unpruned:
+            raise ValidationError(
+                f"Expected {oldest_id} to be in {unpruned!r}, something went wrong during pruning."
+            )
 
-            for depth in range(self._oldest_depth, prune_depth):
-                prune_tasks = tuple(
-                    task_id for task_id in self._tasks.keys()
-                    if self._depths[task_id] == depth
-                )
+    def _validate_has_task(self, task_id: TTaskID) -> TTaskID:
+        if task_id not in self._tasks:
+            raise ValidationError(f"No task {task_id} is present")
+        else:
+            return task_id
+
+    def _find_oldest_unpruned_task_id(self, finished_task_id: TTaskID) -> TTaskID:
+        get_dependency_of_id = compose(
+            self._validate_has_task,
```

Comment:
Looks like this is the reason you're returning the `task_id` from the validate function.  You can use `cytoolz.do` to accomplish that.

```python
from cytoolz.curried import do
get_dependency_of_id = compose(
    do(self._validate_has_task),
    ...
)
```

Reply:
Hah, I just ran across that last week actually. Pretty handy. Until https://github.com/ethereum/eth-utils/issues/136 is done, this is my workaround:

```py
curry(do)(self._validate_has_task)
```